Sources/StreamChat/Utils/Database/NSManagedObject+Extensions.swift

```diff
@@ -163,3 +163,25 @@ extension NSManagedObjectContext {
         return objects
     }
 }
+
+// MARK: - Runtime Filtering
+
+extension Set where Element: NSManagedObject {
+    func filtered(using request: NSFetchRequest<Element>) -> [Element] {
+        guard !isEmpty else { return [] }
+        let result = NSMutableArray(array: (self as NSSet).allObjects, copyItems: false)
+        if let predicate = request.predicate {
+            result.filter(using: predicate)
+        }
+        if let sortDescriptors = request.sortDescriptors, !sortDescriptors.isEmpty {
+            result.sort(using: sortDescriptors)
+        }
+        if request.fetchLimit > 0 {
+            let removeCount = result.count - request.fetchLimit
+            if removeCount > 0 {
+                result.removeObjects(in: NSRange(location: result.count - removeCount, length: removeCount))
+            }
+        }
+        return result as? [Element] ?? []
+    }
```

Comment:
I think we can probably simplify this, and use just regular siltering and sorting like we did with Threads here: https://github.com/GetStream/stream-chat-swift/blob/develop/Sources/StreamChat/Database/DTOs/ThreadDTO.swift#L150-L152

This way it is also consistent with Threads. From now on, this kind of `lastReplies`, `lastMessages` etc... should be filtered and sorted at runtime when converting DTOs to modes as these values usually are very low and it has better performance than performing DB queries.

Reply:
For some of the things we have really complicated filtering. For example, channel messages has 10 predicates. Therefore, I went to reusing the NSPredicate approach because it needs to match with NSFetchRequests passed into DB observers.
But that said, some of the predicates can be skipped, for example, cid predicate because when accessing messages from the ChannelDTO, then I already know these belong to the current channel.

Reply:
https://github.com/GetStream/stream-chat-swift/pull/3325/files#r1686787091

☝️ We should not need this complex predicate here, the backend already returns us the `latestMessages` we don't need to compute it locally. (We did it before but it is not really needed, it is just overengineered)

Reply:
Thank you for the excellent suggestion, this gave another boost, so now we are down to 55 ms from the original 266 ms